### PR TITLE
Feature AppLaunch

### PR DESCRIPTION
**Pull request by LevchenkoS**
Feature allows SDL to launch apps on connected devices.

Related to: [APPLINK-24892](https://adc.luxoft.com/jira/browse/APPLINK-24892)

@okozlovlux, @Kozoriz, @LuxoftAKutsan, @vlantonov, @wolfylambova22, @VVeremjova, please review.

### CODE REVIEW

**Comment by vlantonov** [comment]
@LevchenkoS Is this pull request rebased to the latest commit in develop branch?

**Comment by LevchenkoS** [comment]
@vlantonov, no, it's for release/4.2.0, based on  release/4.1.0

**Comment by vlantonov** [comment]
@LevchenkoS Please try to rebase to the latest commit of release/4.2.0 This pull request contains commits from other people while it is supposed to contain only your commits, related to APPLINK-24892

**Comment by LevchenkoS** [comment]
@vlantonov, sorry, but it should contains them. Please, look at email that I was sent to you and other reviewers.

**Comment by vlantonov** [comment]
@LevchenkoS No, it shouldn't. These commits are from other pull request and are already reviewed and merged. However they appear as changes you have made and thus will be reviewed again. I repeat - please rebase your pull request to the latest commit in release/4.2.0

**Comment by vlantonov** [comment]
@LevchenkoS As far as I understand from your explaination those commits are cherry-picked and that is why they should be here. So this is OK.

File: src/components/application_manager/src/resumption/resume_ctrl.cc [hunk 1]
```diff
@@ -696,7 +701,11 @@ bool ResumeCtrl::CheckDelayAfterIgnOn() {
   return seconds_from_sdl_start <= wait_time;
 }
 
+<<<<<<< HEAD:src/components/application_manager/src/resumption/resume_ctrl.cc
```
Comment (LuxoftAKutsan) [inline src/components/application_manager/src/resumption/resume_ctrl.cc]: @LevchenkoS  remove merge issues

File: src/components/application_manager/src/application_impl.cc [hunk 2]
```diff
@@ -493,7 +493,7 @@ void ApplicationImpl::WakeUpStreaming(
           ServiceType::kMobileNav, true, application_manager_);
```
Comment (LuxoftAKutsan) [inline src/components/application_manager/src/application_impl.cc]: @LevchenkoS Please rework commit message

File: src/components/interfaces/MOBILE_API.xml [hunk 3]
```diff
@@ -1344,30 +1344,6 @@
     </element>
   </enum>
 
-  <struct name="DeviceInfo">
```
Comment (LuxoftAKutsan) [inline src/components/interfaces/MOBILE_API.xml]: @LevchenkoS please squash  this commit with another commit according MOBILE_API

File: src/components/transport_manager/test/transport_manager_impl_test.cc [hunk 4]
```diff
@@ -319,9 +319,7 @@ class TransportManagerImplTest : public ::testing::Test {
   MockTransportAdapter* mock_adapter_;
```
Comment (LuxoftAKutsan) [inline src/components/transport_manager/test/transport_manager_impl_test.cc]: @LevchenkoS Please squash fix commits with Impl commits

File: src/components/application_manager/src/resumption/resume_ctrl.cc [hunk 5]
```diff
@@ -786,7 +795,21 @@ void ResumeCtrl::OnAppRegistrationEnd() {
   StartSavePersistentDataTimer();
 }
 
+<<<<<<< HEAD:src/components/application_manager/src/resumption/resume_ctrl.cc
```
Comment (LuxoftAKutsan) [inline src/components/application_manager/src/resumption/resume_ctrl.cc]: @LevchenkoS remove merge issues

File: src/components/utils/test/test_generator/MOBILE_API.xml [hunk 6]
```diff
@@ -1517,6 +1517,26 @@
```
Comment (LuxoftAKutsan) [inline src/components/utils/test/test_generator/MOBILE_API.xml]: @LevchenkoS Remove MOBILE_API huge changes

File: src/components/application_manager/src/hmi_capabilities_impl.cc [hunk 7]
```diff
@@ -30,181 +30,297 @@
...
-    {"locationDescription",
-     hmi_apis::Common_TextFieldName::locationDescription},
-    {"addressLines", hmi_apis::Common_TextFieldName::turnText},
-    {"turnText", hmi_apis::Common_TextFieldName::addressLines},
-    {"phoneNumber", hmi_apis::Common_TextFieldName::phoneNumber},
-    {"turnText", hmi_apis::Common_TextFieldName::turnText},
-    {"menuTitle", hmi_apis::Common_TextFieldName::menuTitle},
-    {"navigationText", hmi_apis::Common_TextFieldName::navigationText},
-};
+// SDL_CREATE_LOGGER("HMICapabilities");
```
Comment (VVeremjova) [inline src/components/application_manager/src/hmi_capabilities_impl.cc]: @LevchenkoS Commented code

File: src/components/application_manager/src/hmi_capabilities_impl.cc [hunk 8]
```diff
@@ -570,30 +687,125 @@ void HMICapabilities::set_prerecorded_speech(
...
+void HMICapabilitiesImpl::set_phone_call_supported(const bool supported) {
   is_phone_call_supported_ = supported;
 }
 
-void HMICapabilities::Init(resumption::LastState* last_state) {
+void HMICapabilitiesImpl::Init(resumption::LastState* last_state) {
   hmi_language_handler_.Init(last_state);
   if (false == load_capabilities_from_file()) {
-    LOG4CXX_ERROR(logger_, "file hmi_capabilities.json was not loaded");
+    // SDL_ERROR("file hmi_capabilities.json was not loaded");
```
Comment (VVeremjova) [inline src/components/application_manager/src/hmi_capabilities_impl.cc]: @LevchenkoS Commented code

**Comment by wolfylambova22** [comment]
Reviewed.

File: src/components/application_manager/include/application_manager/resumption/resume_ctrl_impl.h [hunk 9]
```diff
@@ -0,0 +1,491 @@
...
+   * @param application is application witch need to be saved
+   */
+  void SaveApplication(app_mngr::ApplicationSharedPtr application) OVERRIDE;
+
+  /**
+   * @brief Set application HMI Level and ausio_state as saved
+   * @param application is application witch HMI Level is need to restore
+   * @return true if success, otherwise return false
+   */
+  bool RestoreAppHMIState(app_mngr::ApplicationSharedPtr application);
```
Comment (vlantonov) [inline src/components/application_manager/include/application_manager/resumption/resume_ctrl_impl.h]: @LevchenkoS Please add OVERRIDE here and in the other methods inherited from `ResumeCtrl`

File: src/components/application_manager/include/application_manager/application_manager_impl.h [hunk 10]
```diff
@@ -839,7 +841,7 @@ class ApplicationManagerImpl
   * @return Resume Controller
   */
   resumption::ResumeCtrl& resume_controller() OVERRIDE {
-    return resume_ctrl_;
+    return *resume_ctrl_.get();
```
Comment (vlantonov) [inline src/components/application_manager/include/application_manager/application_manager_impl.h]: @LevchenkoS Is there any guarantee that `resume_ctrl_` holds a valid pointer?

File: src/components/application_manager/src/application_manager_impl.cc [hunk 11]
```diff
@@ -2122,11 +2145,11 @@ mobile_apis::MOBILE_API& ApplicationManagerImpl::mobile_so_factory() {
 }
 
 HMICapabilities& ApplicationManagerImpl::hmi_capabilities() {
-  return hmi_capabilities_;
+  return *hmi_capabilities_;
```
Comment (vlantonov) [inline src/components/application_manager/src/application_manager_impl.cc]: @LevchenkoS Is there any guarantee that `hmi_capabilities_` is a valid pointer?

File: src/components/application_manager/src/application_manager_impl.cc [hunk 12]
```diff
@@ -2122,11 +2145,11 @@ mobile_apis::MOBILE_API& ApplicationManagerImpl::mobile_so_factory() {
...
 }
 
 const HMICapabilities& ApplicationManagerImpl::hmi_capabilities() const {
-  return hmi_capabilities_;
+  return *hmi_capabilities_;
```
Comment (vlantonov) [inline src/components/application_manager/src/application_manager_impl.cc]: @LevchenkoS Is there any guarantee that `hmi_capabilities_` is a valid pointer?

File: src/components/application_manager/src/application_manager_impl.cc [hunk 13]
```diff
@@ -3459,6 +3482,10 @@ event_engine::EventDispatcher& ApplicationManagerImpl::event_dispatcher() {
   return event_dispatcher_;
 }
 
+app_launch::AppLaunchCtrl& ApplicationManagerImpl::app_launch_ctrl() {
+  return *app_launch_ctrl_;
```
Comment (vlantonov) [inline src/components/application_manager/src/application_manager_impl.cc]: @LevchenkoS Is there any guarantee that `app_launch_ctrl_` is a valid pointer?

File: src/components/application_manager/src/app_launch/apps_launcher.cc [hunk 14]
```diff
@@ -0,0 +1,138 @@
...
+  LauncherPtr app_launcher = *it;
+  works_launchers_.push_back(app_launcher);
+  free_launchers_.erase(it);
+  app_launcher->PosponedLaunch(app_data);
+}
+
+struct AppLauncherFinder {
+  AppLauncherFinder(const ApplicationDataPtr& app_data) : app_data_(app_data) {}
+  bool operator()(const AppsLauncher::LauncherPtr& launcher) const {
+    return *launcher->app_data_ == *app_data_;
```
Comment (vlantonov) [inline src/components/application_manager/src/app_launch/apps_launcher.cc]: @LevchenkoS Is there any guarantee that `app_data_` and `launcher` hold valid pointers?

File: src/components/application_manager/test/hmi_capabilities_test.cc [hunk 15]
```diff
@@ -94,7 +96,7 @@ class HMICapabilitiesTest : public ::testing::Test {
   resumption::LastState last_state_;
   MockApplicationManagerSettings mock_application_manager_settings_;
   utils::SharedPtr<HMICapabilitiesForTesting> hmi_capabilities_test;
-  const std::string kFileName = "hmi_capabilities.json";
+  std::string kFileName;
```
Comment (okozlovlux) [inline src/components/application_manager/test/hmi_capabilities_test.cc]: @LevchenkoS , `file_name`

Comment (LevchenkoS) [inline src/components/application_manager/test/hmi_capabilities_test.cc]: @okozlovlux, fixed in 1834853.

File: src/components/application_manager/include/application_manager/resumption/resume_ctrl_impl.h (hunk 9 above)
Comment (LevchenkoS) [inline src/components/application_manager/include/application_manager/resumption/resume_ctrl_impl.h]: @vlantonov, fixed in c0826c5.

File: src/components/application_manager/src/hmi_capabilities_impl.cc (hunk 8 above)
Comment (LevchenkoS) [inline src/components/application_manager/src/hmi_capabilities_impl.cc]: @VVeremjova, fixed in 1834853.

File: src/components/application_manager/src/hmi_capabilities_impl.cc (hunk 7 above)
Comment (LevchenkoS) [inline src/components/application_manager/src/hmi_capabilities_impl.cc]: @VVeremjova, fixed in 1834853.

File: src/components/utils/test/test_generator/MOBILE_API.xml (hunk 6 above)
Comment (LevchenkoS) [inline src/components/utils/test/test_generator/MOBILE_API.xml]: @LuxoftAKutsan, fixed in e9e372d.

File: src/components/application_manager/src/resumption/resume_ctrl.cc (hunk 5 above)
Comment (LevchenkoS) [inline src/components/application_manager/src/resumption/resume_ctrl.cc]: @LuxoftAKutsan, fixed in c0826c5.

File: src/components/transport_manager/test/transport_manager_impl_test.cc (hunk 4 above)
Comment (LevchenkoS) [inline src/components/transport_manager/test/transport_manager_impl_test.cc]: @LuxoftAKutsan, fixed in c0826c5.

File: src/components/interfaces/MOBILE_API.xml (hunk 3 above)
Comment (LevchenkoS) [inline src/components/interfaces/MOBILE_API.xml]: @LuxoftAKutsan, fixed in e9e372d.

File: src/components/application_manager/src/resumption/resume_ctrl.cc (hunk 1 above)
Comment (LevchenkoS) [inline src/components/application_manager/src/resumption/resume_ctrl.cc]: @LuxoftAKutsan, fixed in c0826c5.

File: src/components/application_manager/src/application_manager_impl.cc (hunk 13 above)
Comment (LevchenkoS) [inline src/components/application_manager/src/application_manager_impl.cc]: @vlantonov, all objects which `ApplicationManagerImpl` is stores, are created within `ApplicationManagerImpl`. So, pointers to `HMICapabilities`, `ResumeCtrl` and `AppLaunchCtrl` are stored inside of it, and initialized by `ApplicationManagerImpl` inside its constructor. And there is no chance that the value of this pointers can be changed from the outside of `ApplicationManagerImpl`.

Comment (vlantonov) [inline src/components/application_manager/src/application_manager_impl.cc]: @LevchenkoS OK, thank you for your answer.

File: src/components/application_manager/src/hmi_capabilities_impl.cc [hunk 16]
```diff
@@ -570,30 +685,125 @@ void HMICapabilities::set_prerecorded_speech(
...
 }
 
-void HMICapabilities::Init(resumption::LastState* last_state) {
+void HMICapabilitiesImpl::Init(resumption::LastState* last_state) {
   hmi_language_handler_.Init(last_state);
   if (false == load_capabilities_from_file()) {
     LOG4CXX_ERROR(logger_, "file hmi_capabilities.json was not loaded");
   } else {
-    LOG4CXX_INFO(logger_, "file hmi_capabilities.json was loaded");
+    LOG4CXX_ERROR(logger_, "file hmi_capabilities.json was loaded");
```
Comment (Kozoriz) [inline src/components/application_manager/src/hmi_capabilities_impl.cc]: Maybe this is just info?

File: src/components/application_manager/test/app_launch/app_launch_data_db_test.cc [hunk 17]
```diff
@@ -0,0 +1,334 @@
...
+
+TEST_F(AppLaunchDataDBTest, SelectMultipleData) {
+  std::vector<ApplicationDataPtr> input_data1;
+  std::vector<ApplicationDataPtr> input_data2;
+  std::vector<ApplicationDataPtr> output_data1;
+  std::vector<ApplicationDataPtr> output_data2;
+  const std::string device_mac_1 = "device_mac_1";
+  const std::string device_mac_2 = "device_mac_2";
+
+  uint32_t half_part = res_db()->get_max_number_iOS_devs() / 2u;
```
Comment (Kozoriz) [inline src/components/application_manager/test/app_launch/app_launch_data_db_test.cc]: Half part of what?

File: src/components/application_manager/test/app_launch/app_launch_data_db_test.cc [hunk 18]
```diff
@@ -0,0 +1,334 @@
...
+  AddApplicationDataWithIncreaseTable(data);
+  recoverData.device_mac_ += "test";
+  GetApplicationData(recoverData, recoverData);
+  EXPECT_FALSE(data == recoverData);
+}
+
+TEST_F(AppLaunchDataDBTest, MaxCount) {
+  const uint32_t mac_ios_devs = res_db()->get_max_number_iOS_devs();
+  for (uint32_t i = 0; i < mac_ios_devs; i++) {
+    ApplicationData data(AddCounter("mobile_app_id_", i),
```
Comment (Kozoriz) [inline src/components/application_manager/test/app_launch/app_launch_data_db_test.cc]: PLease move strings to const fields.

File: src/components/application_manager/test/include/application_manager/mock_app_launch_ctrl.h [hunk 19]
```diff
@@ -0,0 +1,40 @@
...
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+#ifndef SRC_COMPONENTS_APPLICATION_MANAGER_TEST_INCLUDE_APPLICATION_MANAGER_MOCK_APP_LAUNCH_CTRL_H_
+#define SRC_COMPONENTS_APPLICATION_MANAGER_TEST_INCLUDE_APPLICATION_MANAGER_MOCK_APP_LAUNCH_CTRL_H_
+
+#include "application_manager/app_launch/app_launch_ctrl.h"
+namespace test_app_launch {
```
Comment (Kozoriz) [inline src/components/application_manager/test/include/application_manager/mock_app_launch_ctrl.h]: incorrect namespaces 
namespace test ->components ->  app_launch_test

File: src/components/application_manager/test/include/application_manager/mock_app_launch_data.h [hunk 20]
```diff
@@ -0,0 +1,55 @@
...
+
+#ifndef SRC_COMPONENTS_APPLICATION_MANAGER_TEST_INCLUDE_APPLICATION_MANAGER_MOCK_APP_LAUNCH_DATA_H_
+#define SRC_COMPONENTS_APPLICATION_MANAGER_TEST_INCLUDE_APPLICATION_MANAGER_MOCK_APP_LAUNCH_DATA_H_
+
+#include "gmock/gmock.h"
+#include "application_manager/app_launch/app_launch_data.h"
+
+namespace test {
+namespace components {
+namespace test_app_launch {
```
Comment (Kozoriz) [inline src/components/application_manager/test/include/application_manager/mock_app_launch_data.h]: Must be app_launch_test

File: src/components/application_manager/test/include/application_manager/mock_resume_ctrl.h [hunk 21]
```diff
@@ -0,0 +1,100 @@
...
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+#ifndef SRC_COMPONENTS_APPLICATION_MANAGER_TEST_INCLUDE_APPLICATION_MANAGER_MOCK_RESUME_CTRL_H_
+#define SRC_COMPONENTS_APPLICATION_MANAGER_TEST_INCLUDE_APPLICATION_MANAGER_MOCK_RESUME_CTRL_H_
+#include "gmock/gmock.h"
+#include "application_manager/resumption/resume_ctrl.h"
+#include "application_manager/application.h"
+#include "resumption/last_state.h"
+
+namespace resumption {
```
Comment (Kozoriz) [inline src/components/application_manager/test/include/application_manager/mock_resume_ctrl.h]: incorrect namespaces 
test -> components -> resumprion_test

File: src/components/application_manager/test/include/application_manager/mock_app_launch_ctrl.h (hunk 19 above)
Comment (okozlovlux) [inline src/components/application_manager/test/include/application_manager/mock_app_launch_ctrl.h]: @Kozoriz, please add recipient to your comments

File: src/components/include/test/application_manager/mock_app_launch_settings.h [hunk 22]
```diff
@@ -0,0 +1,60 @@
...
+#ifndef SRC_COMPONENTS_INCLUDE_TEST_APPLICATION_MANAGER_MOCK_APP_LAUNCH_SETTINGS_H_
+#define SRC_COMPONENTS_INCLUDE_TEST_APPLICATION_MANAGER_MOCK_APP_LAUNCH_SETTINGS_H_
+
+#include <string>
+#include "gmock/gmock.h"
+#include "application_manager/app_launch_settings.h"
+
+namespace test {
+namespace components {
+namespace test_app_launch {
```
Comment (Kozoriz) [inline src/components/include/test/application_manager/mock_app_launch_settings.h]: must be app_launch_test

File: src/components/application_manager/test/app_launch/app_launch_data_db_test.cc (hunk 18 above)
Comment (LevchenkoS) [inline src/components/application_manager/test/app_launch/app_launch_data_db_test.cc]: @Kozoriz, in this case it's unnecessary. But other strings, which repeats more  frequently has been replaced by constants in 	ec61caa.

File: src/components/include/test/application_manager/mock_app_launch_settings.h (hunk 22 above)
Comment (LevchenkoS) [inline src/components/include/test/application_manager/mock_app_launch_settings.h]: @Kozoriz, fixed in d405378.

File: src/components/application_manager/test/include/application_manager/mock_resume_ctrl.h (hunk 21 above)
Comment (LevchenkoS) [inline src/components/application_manager/test/include/application_manager/mock_resume_ctrl.h]: @Kozoriz, fixed in f6ebfe5.

File: src/components/application_manager/test/include/application_manager/mock_app_launch_data.h (hunk 20 above)
Comment (LevchenkoS) [inline src/components/application_manager/test/include/application_manager/mock_app_launch_data.h]: @Kozoriz, fixed in d405378.

File: src/components/application_manager/test/include/application_manager/mock_app_launch_ctrl.h (hunk 19 above)
Comment (LevchenkoS) [inline src/components/application_manager/test/include/application_manager/mock_app_launch_ctrl.h]: @Kozoriz, fixed in d405378.

File: src/components/application_manager/test/app_launch/app_launch_data_db_test.cc (hunk 17 above)
Comment (LevchenkoS) [inline src/components/application_manager/test/app_launch/app_launch_data_db_test.cc]: @Kozoriz, fixed in d405378.

File: src/components/application_manager/src/hmi_capabilities_impl.cc (hunk 16 above)
Comment (LevchenkoS) [inline src/components/application_manager/src/hmi_capabilities_impl.cc]: @Kozoriz, fixed in 1c2cf9c.

File: src/components/application_manager/src/app_launch/apps_launcher.cc (hunk 14 above)
Comment (LevchenkoS) [inline src/components/application_manager/src/app_launch/apps_launcher.cc]: @vlantonov, fixed in d405378.

**Comment by Kozoriz** [comment]
Reviewed

File: src/components/application_manager/include/application_manager/app_launch/app_launch_ctrl.h [hunk 23]
```diff
@@ -0,0 +1,74 @@
...
+
+#ifndef SRC_COMPONENTS_APPLICATION_MANAGER_INCLUDE_APPLICATION_MANAGER_APP_LAUNCH_APP_LAUNCH_CTRL_H_
+#define SRC_COMPONENTS_APPLICATION_MANAGER_INCLUDE_APPLICATION_MANAGER_APP_LAUNCH_APP_LAUNCH_CTRL_H_
+#include <string>
+#include "utils/shared_ptr.h"
+
+namespace application_manager {
+class Application;
+typedef utils::SharedPtr<const Application> ApplicationConstSharedPtr;
+}  // namesapce application_manager
```
Comment (VVeremjova) [inline src/components/application_manager/include/application_manager/app_launch/app_launch_ctrl.h]: @LevchenkoS typo

**Comment by vlantonov** [comment]
@LevchenkoS Reviewed.

File: src/components/application_manager/test/hmi_capabilities_test.cc [hunk 24]
```diff
@@ -94,7 +98,7 @@ class HMICapabilitiesTest : public ::testing::Test {
   resumption::LastState last_state_;
   MockApplicationManagerSettings mock_application_manager_settings_;
   utils::SharedPtr<HMICapabilitiesForTesting> hmi_capabilities_test;
-  const std::string kFileName = "hmi_capabilities.json";
+  std::string file_name_;
```
Comment (VVeremjova) [inline src/components/application_manager/test/hmi_capabilities_test.cc]: @LevchenkoS You could leave const here

Comment (LevchenkoS) [inline src/components/application_manager/test/hmi_capabilities_test.cc]: @VVeremjova, fixed in 2330691.

File: src/components/application_manager/include/application_manager/app_launch/app_launch_ctrl.h (hunk 23 above)
Comment (LevchenkoS) [inline src/components/application_manager/include/application_manager/app_launch/app_launch_ctrl.h]: @VVeremjova, fixed in 998a28a.

**Comment by okozlovlux** [comment]
@LevchenkoS , reviewed

**Comment by VVeremjova** [comment]
@LevchenkoS Reviewed

File: src/components/include/application_manager/app_launch_settings.h [hunk 25]
```diff
@@ -0,0 +1,22 @@
...
+class AppLaunchSettings {
+ public:
+  virtual const uint16_t app_launch_wait_time() const = 0;
+  virtual const uint16_t app_launch_max_retry_attempt() const = 0;
+  virtual const uint16_t app_launch_retry_wait_time() const = 0;
+  virtual const uint16_t remove_bundle_id_attempts() const = 0;
+  virtual const uint16_t max_number_of_ios_device() const = 0;
+  virtual const uint16_t wait_time_between_apps() const = 0;
+  virtual const bool enable_app_launch_ios() const = 0;
+  virtual uint32_t resumption_delay_after_ign() const = 0;
```
Comment (LuxoftAKutsan) [inline src/components/include/application_manager/app_launch_settings.h]: @LevchenkoS 
either add const to return value or remove const in others

**Comment by LuxoftAKutsan** [comment]
Reviewed

File: src/components/include/application_manager/app_launch_settings.h (hunk 25 above)
Comment (LevchenkoS) [inline src/components/include/application_manager/app_launch_settings.h]: @LuxoftAKutsan, fixed in 9926c88.

**Pull request merged**